### PR TITLE
Implement Mean Absolute Relative Error

### LINKED ISSUE #307: More metrics for regression tasks and forecasting
Following the discussion in #305 
We propose to implement the following list of  metrics

D1 = magnitude of error
- Mean Error
- Mean Normalized Bias
- Fractional Bias

D2 = absolute error
- Median Absolute Error
- Mean Absolute Relative Error / Mean Absolute Percentage Error
- Median Absolute Percentage Error
- Median Relative Absolute Error
- Geometric Mean Absolute Error
- Geometric Mean Relative Absolute Error
- Wave Hedges Distance
- Fractional absolute error
- Canberra Metric

based on the paper: 
https://arxiv.org/pdf/1809.03006.pdf
See appendix 2, page 30.

What do you think ?

Implementation to put into [contrib](https://github.com/pytorch/ignite/tree/master/ignite/contrib/metrics)

### PR DESCRIPTION
**Pull request by weihuangxu**
Fixes #307 

Description:
Add mean absolute relative error to metrics. 

Check list:
* [x] New tests are added (if a new feature is modified)
* [x] New doc strings: text and/or example code are in RST format

### CODE REVIEW

**Comment by vfdev-5** [comment]
@weihuangxu thanks for the PR, could you please put PR template back.

File: ignite/contrib/metrics/MARE.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from __future__ import division
+from ignite.exceptions import NotComputableError
+from ignite.metrics import Metric
+import torch
+
+
+class MeanAbsoluteRelativeError(Metric):
+    """
+    Calculate Mean Absolute Relative Error
+
```
Comment (vfdev-5) [inline ignite/contrib/metrics/MARE.py]: In case of #307 maybe we can put some metrics mathematical definitions and link to the article https://arxiv.org/ftp/arxiv/papers/1809/1809.03006.pdf

**Comment by weihuangxu** [comment]
@vfdev-5 Sure. I have fixed the template. There is an issue regarding the devision in mean absolute relative error. If the ground truth is 0, it may raise numerical issue. I am not sure if we should raise an error or just show Nan in the output. What do you think?

File: ignite/contrib/metrics/MARE.py (hunk 1 above)
Comment (weihuangxu) [reply ignite/contrib/metrics/MARE.py]: That's good idea. I can add equations and link to the paper.

Comment (weihuangxu) [reply ignite/contrib/metrics/MARE.py]: @vfdev-5 It's done. Let me know if there is more modifications needed.

File: ignite/contrib/metrics/MARE.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    def update(self, output):
+        y_pred, y = output
+        absolute_error = torch.abs(y_pred - y.view_as(y_pred)) / torch.abs(y.view_as(y_pred))
+        self._sum_of_absolute_relative_errors += torch.sum(absolute_error).item()
+        self._num_samples += y.shape[0]
+
+    def compute(self):
+        if self._num_samples == 0:
+            raise NotComputableError('MeanAbsoluteError must have at least one example before it can be computed')
```
Comment (TheCodez) [inline ignite/contrib/metrics/MARE.py]: Should be `MeanAbsoluteRelativeError` instead.

**Comment by vfdev-5** [comment]
@weihuangxu maybe we could add a flag ~ `percent=True` to return Mean Absolute Percentage Error = MARE * 100 ?

> There is an issue regarding the devision in mean absolute relative error. If the ground truth is 0, it may raise numerical issue. I am not sure if we should raise an error or just show Nan in the output. What do you think?

@weihuangxu good question, not sure how to regularize this properly. 
Maybe we can add something like:
```
a = 0
if error < tol (1e-20):
     MARE = 0
else:
     MARE = Inf
``` 
@zasdfgbnm any ideas ?

File: ignite/contrib/metrics/MARE.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/contrib/metrics/MARE.py]: @TheCodez thank you :)

Comment (vfdev-5) [inline ignite/contrib/metrics/MARE.py]:
```suggestion
            raise NotComputableError('MeanAbsoluteRelativeError must have at least one example before it can be computed')
```

**Comment by zasdfgbnm** [comment]
@vfdev-5 Considering metric arithmetic (https://github.com/pytorch/ignite/issues/321) is coming soon, would it be helpful to complicate the API by adding a `percent` argument? Builtin metrics like `Precision` does not have such thing.

For the divide by zero error, I would suggest just ignore the case that ground truth is zero, doing so the final result would automatically becomes `inf`,  and the user should know  preprocess their dataset if they want MARE.

**Comment by vfdev-5** [comment]
@zasdfgbnm good point with `percent`. Ok, let's skip this flag.

> For the divide by zero error, I would suggest just ignore the case that ground truth is zero, doing so the final result would automatically becomes inf, and the user should know preprocess their dataset if they want MARE.

So you suggest to ignore the cases with gt = 0 ? 

I did not get your point about "the user should know preprocess their dataset if they want MARE." ...

**Comment by zasdfgbnm** [comment]
@vfdev-5 as from the definition, MARE = sum(|error/gt|), gt is something given in the dataset, which I assume the users should but might not know well. If something in the dataset has gt=0, why do they insist on using ill-defined metrics like MARE? So all we need to do is to notify the user that they should not use this metric, by returning a `inf` or `nan` or raise an Exception.

So, considering that if we don't add any special handle of case gt=0, the user would know that they are doing things wrongly by getting a `inf` or `nan` result. But if we add 
```python
if error < tol (1e-20):
     MARE = 0
else:
     MARE = Inf
```
the user might sometimes get 0 instead of `nan`, which I consider as a silent failure.

**Comment by vfdev-5** [comment]
@zasdfgbnm Ok, I see your point. Let's raise an exception `NotComputableError` with a message that the ground truth is zero.

**Comment by weihuangxu** [comment]
@TheCodez Thanks for pointing out. 

@vfdev-5 I added the error message in update step when the ground truth has 0 value.

**Comment by vfdev-5** [comment]
@weihuangxu thanks! Some comments in the code and could you also please check if docs are correctly rendered with formula:
```
cd docs
pip install -r docs/requirements.txt
make html
cd build && python3 -m http.server 1234
```
and go in the browser to 0.0.0.0:1234

File: ignite/contrib/metrics/MARE.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    """
+
+    def reset(self):
+        self._sum_of_absolute_relative_errors = 0.0
+        self._num_samples = 0
+
+    def update(self, output):
+        y_pred, y = output
+        if (y == 0).any():
+            raise NotComputableError('The ground truth has 0')
```
Comment (vfdev-5) [inline ignite/contrib/metrics/MARE.py]: Could you please also add a test on this

Comment (weihuangxu) [reply ignite/contrib/metrics/MARE.py]: Sure. I will work on this soon.

**Comment by weihuangxu** [comment]
@vfdev-5  Thanks for reviewing the code. Here are a couple modifications:

- Add the test for raising error for zero division.

- Add the test for raising error for zero sample.

- Use random values instead of fixed values in test_MARE.

- Change the reference name to Botchkarev 2018 and successfully build the docs.

File: ignite/contrib/metrics/MARE.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+from __future__ import division
```
Comment (vfdev-5) [inline ignite/contrib/metrics/MARE.py]: nit: maybe it'd be better to rename the file as `mean_absolute_relative_error` instead of `MARE`

File: ignite/contrib/metrics/test_MARE.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
+from ignite.contrib.metrics import MeanAbsoluteRelativeError
```
Comment (vfdev-5) [inline ignite/contrib/metrics/test_MARE.py]: Test file need to be with all other test files :)

File: ignite/contrib/metrics/MARE.py (hunk 4 above)
Comment (weihuangxu) [reply ignite/contrib/metrics/MARE.py]: Sure. It is important to keep consistent. If other metrics use full name, it would be better to change it.

File: ignite/contrib/metrics/test_MARE.py (hunk 5 above)
Comment (weihuangxu) [reply ignite/contrib/metrics/test_MARE.py]: Fixed.

File: ignite/contrib/metrics/MARE.py (hunk 4 above)
Comment (weihuangxu) [reply ignite/contrib/metrics/MARE.py]: Fixed.

File: tests/ignite/contrib/metrics/test_mean_absolute_relative_error.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    sum_samples += num_samples_d
+    MARE_d = sum_error / sum_samples
+    assert m.compute() == approx(MARE_d.item())
+
+
+def test_zero_div():
+    a = torch.tensor([2.0, -1.0, -1.0, 2.0])
+    ground_truth = torch.tensor([0.0, 0.5, 0.2, 1.0])
+
+    m = m = MeanAbsoluteRelativeError()
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/test_mean_absolute_relative_error.py]:
```suggestion
    m = MeanAbsoluteRelativeError()
```

File: tests/ignite/contrib/metrics/test_mean_absolute_relative_error.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+def test_mean_absolute_relative_error():
+    a = torch.rand(4)
+    b = torch.rand(4)
+    c = torch.rand(4)
+    d = torch.rand(4)
+    ground_truth = torch.rand(4)
+
+    m = MeanAbsoluteRelativeError()
+    m.reset()
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/test_mean_absolute_relative_error.py]: I was discussed in some of previously proposed PRs that `reset` is not needed after the constructor. We can remove.

File: tests/ignite/contrib/metrics/test_mean_absolute_relative_error.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    MARE_d = sum_error / sum_samples
+    assert m.compute() == approx(MARE_d.item())
+
+
+def test_zero_div():
+    a = torch.tensor([2.0, -1.0, -1.0, 2.0])
+    ground_truth = torch.tensor([0.0, 0.5, 0.2, 1.0])
+
+    m = m = MeanAbsoluteRelativeError()
+    m.reset()
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/test_mean_absolute_relative_error.py]: Same here. No need to reset

File: tests/ignite/contrib/metrics/test_mean_absolute_relative_error.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    with raises(NotComputableError):
+        m.update((a, ground_truth))
+
+
+def test_zero_sample():
+    m = MeanAbsoluteRelativeError()
+    m.reset()
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/test_mean_absolute_relative_error.py]: Here too, reset is useless

File: ignite/contrib/metrics/mean_absolute_relative_error.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+class MeanAbsoluteRelativeError(Metric):
+    r"""
+    Calculate Mean Absolute Relative Error
+
+    :math:`\text{MARE} = \frac{1}{n}\sum _j^n\frac{\left|A_j-P_j\right|}{\left|A_j\right|}`
+
+    where, :math:`A_j` is the ground truth and :math:`P_j` is the predicted value.
+
+    More details can be found in the reference `Botchkarev 2018`_ .
```
Comment (zasdfgbnm) [inline ignite/contrib/metrics/mean_absolute_relative_error.py]: What is the difference between referencing as 
```rst
`text`__
__ https://link.to/your/target
```

vs 

```rst
`text`_
.. _ `text`:
    https://link.to/your/target
```
?
If we use the latter one for different metrics but with same text and url, would it create errors like in https://github.com/pytorch/ignite/pull/331#issuecomment-438177435
